Routing/multiple table ranges (#5432)
projectcalico/calico: calico/reference/resources/felixconfig.md

| reportingInterval                  | Interval at which Felix reports its status into the datastore, or 0 to disable.  Must be non-zero in OpenStack deployments. | `5s`, `10s`, `1m` etc. | duration | `30s` |
| reportingTTL                       | Time-to-live setting for process-wide status reports. | `5s`, `10s`, `1m` etc. | duration | `90s` |
| routeRefreshInterval               | Period at which Felix re-checks the routes in the dataplane to ensure that no other process has accidentally broken {{site.prodname}}'s rules. Set to 0 to disable route refresh. | `5s`, `10s`, `1m` etc. | duration | `90s` |

> | routeTableRange                    | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes.  `RouteTableRange` specifies the indices of the route tables that Calico should use. |  | [RouteTableRanges](#routetablerange) | `{Min: 1, Max: 250}` |

Reviewer: Same here, should the default be unset?

> | routeTableRange                    | *deprecated in favour of `RouteTableRanges`* Calico programs additional Linux route tables for various purposes. `RouteTableRange` specifies the indices of the route tables that Calico should use. If explicitly-set, will be honoured in favour of `RouteTableRanges`. |  | [RouteTableRanges](#routetablerange) | `{Min: 1, Max: 250}` |

Reviewer: Left the default on this one.

Author: good catch, thought I got them all :laughing:

Reviewer 2: Warning: American incoming

"honoured -> honored"

"favour" -> "favor" 

alternatively, we can avoid those contentious words altogether :laughing: 

> If explicitly set, this field will take precendence over RouteTableRanges

Non-american comment: "explicitly-set" shouldn't be hyphenated. "explicitly set" is fine.

Reviewer 3: s/precendence/precedence

Author: cleaned up this wording now, although worth noting I saw a few unrelated parts of our codebase that still use european spelling, I dont know if a formal decision has been made around stuff like this (?), but maybe another, more intentional PR could grab those remaining bits if we want to make a concerted effort there

| protocol | The protocol match   | tcp, udp, sctp                       | string |
| net      | The CIDR match       | any valid CIDR (e.g. 192.168.0.0/16) | string |

#### RouteTableRanges
`RouteTableRanges` is a list of `RouteTableRange` objects:

| Field    | Description          | Schema |
|----------|----------------------|--------|
| min      | Minimum index to use | int    |
| max      | Maximum index to use | int    |

> Each `RouteTableRange` designates a set of routing tables available to Calico. Modern Linux kernels can theorectically support more than the documented 252 additional tables. However, allocating more than the default number of tables is not as simple as increasing the max table index, since the increased range would encpomass the reserved tables, and could lead to Calico clashing with the kernel.

Reviewer 4: Typo in `encpomass`.

Author: nice catch!

| protocol | The protocol match   | tcp, udp, sctp                       | string |
| net      | The CIDR match       | any valid CIDR (e.g. 192.168.0.0/16) | string |

#### RouteTableRanges
`RouteTableRanges` is a list of `RouteTableRange` objects:

| Field    | Description          | Schema |
|----------|----------------------|--------|
| min      | Minimum index to use | int    |
| max      | Maximum index to use | int    |

> Each `RouteTableRange` designates a set of routing tables available to Calico. Modern Linux kernels can theorectically support more than the documented 252 additional tables. However, allocating more than the default number of tables is not as simple as increasing the max table index, since the increased range would encompass the reserved tables, and could lead to Calico clashing with the kernel.

Reviewer: Can live with this approach (we've got lots of config parameters that are similarly handled), but (with experience gained from all those parameters...) I think it'd be better for the user if you just automatically excluded the reserved Linux tables.  I.e. let the user specify `10-10000` and just discard the `25x` tables with an INFO log `"Route table range contains linux reserved range, excluding 253-255."`.  

Then I'd include a note in the docs that Felix automatically excludes `0 and 253-255` since they are reserved by Linux to let knowledgeable users know so they don't freak out.   Rule of thumb: if the explanation in the docs can be shorter/easier, it might be a good idea.

Author: That does sound like a favourable solution in fairness! I think the main reason I opted for this approach is that different provisioners will have different reserved tables - E.g, in the case of AWS, as noted in the docs. I'm not hugely familiar with how much variance there is across provisioners, but maintaining a list of which ones reserve which tables gave me a hunch it may be better to go with the verbose option.
Totally makes sense what you're saying though, and my worries may be overblown! :slightly_smiling_face:

Reviewer: I was only thinking of ignoring the Linux-reserved tables.  Will still need config for other provisioners.

| min      | Minimum index to use | int    |
| max      | Maximum index to use | int    |

Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-10000`, and will automatically ignore Linux's default reserved tables `253-255`. However, it's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables below and above the sensitive ranges:
```sh
# target tables 65-99, and 256-1000, skipping 100-255
calicoctl patch felixconfig default --type=merge -p '{"spec":{"routeTableRanges": [{Min: 65, Max: 99}, {Min: 256, Max: 10000}] }}
```

*Note*, for performance reasons, the maximum total number of routing tables that Felix will accept is 65535 (or 2*16).

> If the deprecated `RouteTableRange` option is explicitly set, it will be honoured in favour of `RouteTableRanges`.

Reviewer: I think I'd have jumped the other way if both are set.  If both are explicitly set, I'd use the value from RouteTableRanges.

Author: I initially wanted to go that route, but I encountered a problem:
 - remove `RouteTableRange` default so we can check when it's empty
 - `RouteTableRanges` still has a non-zero default, so it will always look "set"-ish
 - if the user the deprecated arg and left the new one as default, while I _could_ check that `RouteTableRanges` arg isnt _default_, and go with the deprecated arg, there would be ambiguity - if they for some reason explicitly set `RouteTableRanges` to the default val, I wouldn't be able to tell that apart :thinking: 
 
I didn't immediately see a way from within the config function to differentiate between set (to the default) and unset default, so I opted to give the old param precedence when it's there, since if it's unset, it'll definitely be a zero-val

Hope that makes sense. maybe Im missing a way around this?

Reviewer: One option would be to remove both their defaults in the Felix config_params.go and then do the defaulting "manually" in the getter function.  We do it that way to handle a couple of other "more complicated" defaults.

Author: Thats a great suggestion :raised_hands: will implement.

Reviewer 2: I believe the way that Kubernetes has handled API changes from singular to plural is through validation that both are set consistently.

e.g., if `RouteTableRange` and `RouteTableRanges` are set, they need to be consistent with each other else they will error. 

That might just mean "Return an error on the API if both are set" - that seems like the cleanest behavior to me.

Reviewer 2: The way I think about this is that the code shouldn't be assuming or trying to guess the users intent - if the user gives us two pieces of conflicting intent via the API, we need to tell the user "Hey, you're not making sense".

Author: I like the suggestion, personally am happy with either approach. Are there other examples of either solution being done before in our code? If this is the first occurence, I'm happy to put the best foot forward, but Im also conscious of getting choice paralysis here as I'd be rewriting this piece of the PR for the third time :laughing:  if it's something that we're ambiguous about, I might acknowledge the suggestion but leave it off this time around. WDYT?

Reviewer 2: I think the issue with FelixConfiguration that makes my suggestion above not entirely viable is that the config that Felix ultimately sees might have both of these set, even if we prevent them from both being set on a single API object, since Felix merges config from multiple API objects as well as env vars, files, etc. So we need to think of this in two ways: What does the API do with this, and what does Felix do with this.

I think we should do effectively what Shaun suggested above, but with a little bit of API validation so that on a single object only one of these is ever set. i.e.,

- The API validation code should prevent both of these from being set on the same object (libcalico-go validation func) 
- Remove Felix config_param defaulting per @fasaxc's suggestion above. 
- Felix code should emit a warning log if it sees both config options as being set.
- Felix code should emit a warning log if it sees the old, deprecated option as being set.
- Felix code prefers the new option if both are set.

I think that means you probably don't need to re-work this logic again.

Author: Sounds reasonable! Will implement :smiley: thanks guys

> | min      | Minimum index to use | int    |

Reviewer: Add the Accepted values column here?

| max      | Maximum index to use | int    |

> Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-250`, and will automatically ignore Linux's default reserved tables `253-255`. However, it's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables below and above the sensitive ranges:

Reviewer: ```suggestion
Each item in the `RouteTableRanges` list designates a range of routing tables available to Calico. By default, Calico will use a single range of `1-250`.  If a range spans Linux's reserved table range (`253-255`) then those tables are automatically excluded from the list. It's possible that other table ranges may also be reserved by third-party systems unknown to Calico. In that case, multiple ranges can be defined to target tables below and above the sensitive ranges:
```

```sh
# target tables 65-99, and 256-1000, skipping 100-255
calicoctl patch felixconfig default --type=merge -p '{"spec":{"routeTableRanges": [{"Min": 65, "Max": 99}, {"Min": 256, "Max": 1000}] }}
```

*Note*, for performance reasons, the maximum total number of routing tables that Felix will accept is 65535 (or 2*16).

> `RouteTableRanges` will override `RouteTableRange` if both are set.

Reviewer: ```suggestion
If both `RouteTableRanges` and `RouteTableRange` are set, `RouteTableRanges` takes precedence and `RouteTableRange` is ignored.
```
